### PR TITLE
Revise pass_by semantics for port XML

### PR DESCRIPTION
**Pull request by bocchino**
| | |
|:---|:---|
|**_Originating Project/Creator_**| F Prime infrastructure |
|**_Affected Component_**| Various |
|**_Affected Architectures(s)_**| Various |
|**_Related Issue(s)_**| N/A |
|**_Has Unit Tests (y/n)_**| y |
|**_Builds Without Errors (y/n)_**| y |
|**_Unit Tests Pass (y/n)_**| y |
|**_Documentation Included (y/n)_**| y |

---
## Change Description

This PR revises the behavior of the `pass_by` attribute in the port XML as follows:

* Remove `pass_by="value"` from the schema.
* Revise the default behavior (in the case of no `pass_by` attribute) as follows: (1) For arguments of primitive type, pass by value. (2) For arguments of non-primitive type, pass by const reference at the C++ level. At the F Prime level, (2) is an actual pass by const reference across a synchronous port invocation. Across an asynchronous port invocation, it means (a) make a local copy of the data and (b) pass a const reference to the local copy.

This PR revises port usage as necessary (e.g.., remove the `value` keyword, replace pass-by-value with pass-by-const-reference in handler implementations).

## Rationale

This change simplifies the XML semantics and is consistent with FPP. It removes a source of errors where large values could inadvertently be passed by value across several internal function calls.

## Testing/Review Recommendations

This PR affects a basic framework behavior, so it should get a careful review. One might start with the new unit test here: https://github.com/bocchino/fprime/tree/pass-by/Autocoders/Python/test/pass_by_kind. This test covers the new behavior.

## Future Work

There is one more PR to go to integrate the summer work from @rohandkn. It has to do with differentiating array lengths from string sizes.

### CODE REVIEW

**Comment by bocchino** [comment]
It looks like there is an issue with the FPP installation on the CI platform. I'm looking into it.

**Comment by LeStarch** [comment]
@bocchino what are the update recommendations for users who are currently using pass_by="value" to maintain similar behavior?

File: Autocoders/Python/test/enum_xml/main.cpp [hunk 1]
```diff
@@ -17,25 +17,21 @@
 
 using namespace std;
 
-// Component instance pointers
-Example::ExampleEnumImpl* inst1 = nullptr;
-Example::ExampleEnumImpl* inst2 = nullptr;
+// Instantiate the inst1 and inst2 components
```
Comment (LeStarch) [inline Autocoders/Python/test/enum_xml/main.cpp:20]: Why was this changed?  General cleanup?

File: Autocoders/Python/src/fprime_ac/utils/TypesList.py [hunk 2]
```diff
@@ -1,19 +1,19 @@
 # types allowed in commands/telemetry/parameters/events
 types_list = [
-    "U8",
+    "bool",
```
Comment (LeStarch) [inline Autocoders/Python/src/fprime_ac/utils/TypesList.py:3]: why was this changed?

Comment (LeStarch) [reply Autocoders/Python/src/fprime_ac/utils/TypesList.py:3]: Alphabetical order?

**Comment by bocchino** [comment]
> what are the update recommendations for users who are currently using pass_by="value" to maintain similar behavior?

1. Remove `pass_by="value"` everywhere.
2. Revise your derived handlers to use const T& instead of T, for non-primitive types that were passed by value.
2. In most cases, steps 1 and 2 should be enough. If you need a copy of the data, instead of the reference, make a copy, e.g.:
```
void port_handler(const T& a) {
  // Copy the underlying data from a to a1
  auto a1 = a;
  // use a1 here
}

File: Svc/ActiveLogger/ActiveLoggerImpl.cpp [hunk 3]
```diff
@@ -100,8 +100,10 @@ namespace Svc {
             }
         }
 
+        Fw::LogSeverity severity_copy = severity;
+
         // send event to the logger thread
-        this->loqQueue_internalInterfaceInvoke(id,timeTag,severity,args);
+        this->loqQueue_internalInterfaceInvoke(id,timeTag,severity_copy,args);
```
[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp:106]: This inconsistency is woven into the autocoder, via similar-but-not-quite-the-same code for handling ports and internal interfaces. The current PR does not change the behavior for internal interfaces.

I think the current behavior for internal interfaces is acceptable. It ignores the `pass_by` attribute and unconditionally passes primitive types by value and non-primitive types by (mutable) reference.

I recommend that we accept this behavior for release 3.0 and revisit later, when we revise the entire C++ back end.

[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp:106]: I removed `pass_by="reference"` from the internal interfaces of the active loggers, because it does nothing.

File: Autocoders/Python/test/enum_xml/main.cpp (hunk 1 above)
Comment (bocchino) [reply Autocoders/Python/test/enum_xml/main.cpp:20]: Yes, I believe we used this test as the basis for a new test, noticed that this test used an older style for creating test components, and modernized both tests.

File: Autocoders/Python/src/fprime_ac/utils/TypesList.py (hunk 2 above)
Comment (bocchino) [reply Autocoders/Python/src/fprime_ac/utils/TypesList.py:3]: Yes, we sorted the list.

File: Svc/ActiveLogger/ActiveLoggerImpl.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp:106]: However, I don't know why `severity_copy` was added there. I will investigate.

[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp:106]: Actually, now that I look at the diff in ActiveLogger, it seems that before this PR the autocoder was sensitive to `pass_by="reference"`. I'll have to take another look at the autocoder code.

[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp:106]: I guess `severity_copy` was added in order to pass the data from a const ref into a non-const ref. As I mentioned, we could fix this but it would take some work. I think it's out of scope for this PR.

[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp:106]: Possibly we could revise the internal interface code to use a const ref instead of a non-const ref. Maybe we could do this without major changes. I will investigate.

[the remark this answers is not in the document]
Comment (bocchino) [reply Svc/ActiveLogger/ActiveLoggerImpl.cpp]: Confirmed, I improved the const correctness and removed the extra copy. Thanks, @LeStarch!

**Comment by bocchino** [comment]
In light of what we learned here, I'm disallowing `ref` parameters in FPP internal ports. See https://github.com/fprime-community/fpp/issues/95.

**Review by LeStarch — APPROVED** [review]
No further comments.
